Improve typing and docstrings

felixblanke opened the pull request:
This addresses #85 and changes type hints throughout the code base.

Notably, I added two type aliases as return types for wavelet transforms:
1. `WaveletCoeffNd` which corresponds to  `(Tensor, *(WaveletDetailDict, ...))` 
2. `WaveletCoeff2d` which corresponds to `(Tensor, *(WaveletDetailTuple2d, ...))` 

which make use of the following two new type aliases for detail coefficients:
1. `WaveletDetailDict` which corresponds to `dict[str, Tensor]`
2. `WaveletDetailTuple2d` which corresponds to `(Tensor, Tensor, Tensor)`.

I also added the type alias `WaveletCoeff2dSeparable` which is an alias for `WaveletCoeffNd` to emphasize the different return type in case of the 2d separable transform

_(I would be very much open to suggestions how to name those differently)_

Importantly: The collection that contains the wavelet coefficients changes from `list` to `tuple`. Therefore, **this changes the return types of our wavelet coefficients**, not "only" for type hinting but also with possible impact on user code (e.g. we change the collection from mutable to immutable). 

Further changes:
* Some typing is narrowed (e.g. by replacing list by sequence)
* For typing we switch from `typing.List` to the builtin `list` (and the same for `dict` and `tuple`). This was introduced in python 3.9 (our minimum version)
* I adopted the [recommended flake8 config](https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html#flake8) by the black project to address differing opinions between both projects.
* Minor improvements to the docs and docstrings

**Important note for checking the type hints:**
The overloads for `_map_result` introduced in the `_util` module currently lead to a crash in mypy ([relevant issue](https://github.com/python/mypy/issues/17319)). Thus, running the nox typing session will fail.

**Update:** A fix has been merged to mypy's master branch ([relevant PR](https://github.com/python/mypy/pull/17356)). To install from the git repository, replace in `setup.cfg:70` the line `mypy` by
```
    mypy @ git+https://github.com/python/mypy
```

v0lta commented:
Thanks a lot, @felixblanke. Since it's older, I will start with https://github.com/v0lta/PyTorch-Wavelet-Toolbox/pull/84 and then move on to this PR.

cthoyt commented on src/ptwt/_stationary_transform.py:
```diff
@@ -28,7 +29,7 @@ def _swt(
 
     Returns:
-        List[torch.Tensor]: Same as wavedec.
+        list[torch.Tensor]: Same as wavedec.
```
as you go through these, it's better to remove the redundant type annotations inside the docstrings rather than update them

```suggestion
        : Same as wavedec.
````

felixblanke replied:
Should we keep the leading colons? The examples in [Google's styleguide](https://google.github.io/styleguide/pyguide.html#doc-function-returns) show none.

cthoyt replied:
if the documentation works without it, then no! I am not usually using this code style so I don't know all the specifics

felixblanke replied:
I went through the docstrings and cleaned them up. This should be addressed now.

The styleguide also writes about function arguments: "The description should include required type(s) if the code does not contain a corresponding type annotation." Opinions on this matter?

cthoyt commented on src/ptwt/_util.py:
```diff
@@ -29,6 +31,15 @@ def __len__(self) -> int:
         return len(self.dec_lo)
 
 
+WaveletDetailTuple2d = tuple[torch.Tensor, torch.Tensor, torch.Tensor]
```
it's useful to add a docstring to each of these tuples to give some additional context about what they are and where they're used

cthoyt replied:
also, this could be a place to use a namedtuple to make it even more explicit

felixblanke replied:
Thats a good idea :+1:

cthoyt commented on src/ptwt/_util.py:
```diff
@@ -137,49 +148,75 @@ def _check_if_tensor(array: Any) -> torch.Tensor:
             result_lst.append(new_dict)
-    return result_lst
+        else:
+            raise AssertionError(f"Unexpected input type {type(element)}")
```
```suggestion
            raise ValueError(f"Unexpected input type {type(element)}")
```

cthoyt commented on src/ptwt/conv_transform_2.py:
```diff
@@ -214,51 +212,52 @@ def wavedec2(
+
     result_lst.reverse()
+    res_ll = res_ll.squeeze(1)
+    result: WaveletCoeffDetailTuple2d = res_ll, *result_lst
```
might be an opportunity to rethink this data structure to have more nested parts rather than a variable length part

felixblanke replied:
I agree that such a data structure would preferable.
However, the current data structure is chosen to follow the conventions of `pywt` (which expects a tuple or list of the current form, see [the API reference](https://pywavelets.readthedocs.io/en/latest/ref/2d-dwt-and-idwt.html#pywt.waverec2)). Also, moving to more nested parts would probably break significantly more user code.

cthoyt replied:
in that case, better to stick to what's already here, but just leave a comment on the definition of the data structure that this is the reason :)

felixblanke commented:
I just noticed that that the type aliases like `WaveletCoeffDetailTuple2d` are not properly displayed in the docs when I build them. I will look into that tomorrow

v0lta commented on src/ptwt/_stationary_transform.py:
```diff
@@ -18,18 +19,18 @@ def _swt(
-    axis: Optional[int] = -1,
-) -> List[torch.Tensor]:
+    axis: int = -1,
+) -> list[torch.Tensor]:
```
This has worked since Python 3.9 ( https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections ). We do not support 3.8, it is okay. Thanks for updating!

v0lta commented on src/ptwt/_util.py:
```diff
@@ -29,6 +31,15 @@ def __len__(self) -> int:
+WaveletCoeffDetailTuple2d = tuple[
+    torch.Tensor, Unpack[tuple[WaveletDetailTuple2d, ...]]
+]
+WaveletCoeffDetailDict = tuple[torch.Tensor, Unpack[tuple[WaveletDetailDict, ...]]]
```
This appears in the documentation, no? Should we perhaps call it 3D-wavelet details coeffs? After all, it's called Dict now, but it's actually a tuple with dicts. This is the structure we inherit from 'pywt'; perhaps dropping the dict-type from the name reduces the risk of confusion.

felixblanke replied:
I find the naming to be quite hard :D I agree, having the `Dict` / `Tuple2d` in the name invokes the idea that the object itself is of this type. However, I didn't come up with a good way to put the difference between both types into the names. I thought about adding the dimensions (i.e. `Tuple2d -> 2d`, `Dict -> Nd`) but we also use the dict version in the separable 2d case.

v0lta replied:
Yes, I agree. Could we create an alias for the separable 2d case? Would having both a separable coeffs and an Nd coeffs type be a solution?

felixblanke replied:
I renamed
* `WaveletCoeffDetailDict` to `WaveletCoeffNd`
* `WaveletCoeffDetailTuple2d` to `WaveletCoeff2d`
and introduced
* `WaveletCoeff2dSeparable` as an alias for `WaveletCoeffNd`
but kept the names for the detail coefficient type aliases.
Do you think that works?

v0lta replied:
Yes! I think it does.

felixblanke commented:
This PR changes the returned collection for higher dimensional wavelet transforms from list to tuple to enable better type hints. This is not done for the 1d case where we still return `list[Tensor]`.
Should we change this to `tuple[Tensor, ...]` to stay consistent with the higher dimensional transforms? Or should we just leave it as is? @v0lta

v0lta commented:
https://pywavelets.readthedocs.io/en/latest/ref/dwt-discrete-wavelet-transform.html#multilevel-decomposition-using-wavedec also returns a list, although consistency would, of course, be nice. While I lean towards leaving things as is, I would accept the change for consistency. However, if someone comes from `pywt` and has code that builds in the list type, this could cause trouble. On the other hand, that's unlikely to happen. Hence, I am just leaning toward letting it be how it is.

felixblanke commented:
I also went through the examples. In consequence, I refactored the functions  creating the natural and frequency packet order for `WaveletPacket2D`.
I changed `get_natural_order` and `get_frequency_order` both to be static functions of `WaveletPacket2D`.
Further, to make both methods consistent, `get_frequency_order` now returns concatenated strings instead of a tuple of single char strings.

v0lta commented:
do we add ` mypy @ git+https://github.com/python/mypy ` to setup.py until the next  mypy release?

v0lta commented:
okay. I think we are done here. Thanks a lot @felixblanke and @cthoyt . I will merge this and close https://github.com/v0lta/PyTorch-Wavelet-Toolbox/issues/85 . Feel free to reopen if we are still missing something.
